xline/src/server/maintenance.rs

```diff
@@ -0,0 +1,225 @@
+                // and the check method size % 512 == sha256.size, So we need to pad snapshots to multiples
+                // of 512 bytes
+                let padding = 512.overflow_sub(buf_size.overflow_rem(512));
+                if padding != 0 {
+                    buf.append(&mut vec![0; padding.cast()]);
+                }
+                hasher.update(&buf);
+                let resp: SnapshotResponse = SnapshotResponse {
+                    header: Some(header.clone()),
+                    remaining_bytes: remain_size,
+                    blob: buf,
+                };
+                if let Err(e) = tx.send(Ok(resp)).await {
+                    error!("snapshot send failed, {e}");
+                    return;
+                }
+            }
+            let sha = hasher.finalize().to_vec();
+            let resp = SnapshotResponse {
+                header: Some(header),
+                remaining_bytes: 0,
+                blob: sha,
+            };
+            if let Err(e) = tx.send(Ok(resp)).await {
+                error!("snapshot send failed, {e}");
+                return;
+            }
+            if let Err(e) = snapshot.clean().await {
+                error!("snapshot clean failed, {e}");
+            }
+        });
+
+        Ok(tonic::Response::new(ReceiverStream::new(rx)))
+    }
+
+    async fn move_leader(
+        &self,
+        _request: tonic::Request<MoveLeaderRequest>,
+    ) -> Result<tonic::Response<MoveLeaderResponse>, tonic::Status> {
+        Err(tonic::Status::unimplemented(
+            "move_leader is unimplemented".to_owned(),
+        ))
+    }
+
+    async fn downgrade(
+        &self,
+        _request: tonic::Request<DowngradeRequest>,
+    ) -> Result<tonic::Response<DowngradeResponse>, tonic::Status> {
+        Err(tonic::Status::unimplemented(
+            "downgrade is unimplemented".to_owned(),
+        ))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::{error::Error, path::PathBuf};
+
+    use tokio_stream::StreamExt;
+    use utils::config::StorageConfig;
+
+    use super::*;
+    use crate::storage::db::DBProxy;
+
+    #[tokio::test]
+    async fn test_snapshot_rpc() -> Result<(), Box<dyn Error>> {
+        let dir = PathBuf::from("/tmp/test_snapshot_rpc");
+        let db_path = dir.join("db");
+
+        let persistent = DBProxy::open(&StorageConfig::RocksDB(db_path.clone()))?;
+        let header_gen = Arc::new(HeaderGenerator::new(0, 0));
+        let maintenance_server = MaintenanceServer::new(persistent, header_gen);
+        let mut snap1_stream = maintenance_server
+            .snapshot(tonic::Request::new(SnapshotRequest {}))
+            .await?
+            .into_inner();
+        let mut recv_data = Vec::new();
+        while let Some(data) = snap1_stream.next().await {
+            recv_data.append(&mut data?.blob);
+        }
+        assert_eq!(recv_data.len() % 512, 32);
```

Comment:
Why is 32 here?

Reply:
size of `sha256`, refactored in a cleaner way